packages/three-vrm/src/VRMUtils/combineSkeletons.ts

```diff
@@ -0,0 +1,115 @@
+import * as THREE from 'three';
+
+/**
+ * Traverses the given object and combines the skeletons of skinned meshes.
+ *
+ * Each frame the bone matrices are computed for every skeleton. Combining skeletons
+ * reduces the number of calculations needed, improving performance.
+ *
+ * @param root Root object that will be traversed
+ */
+export function combineSkeletons(root: THREE.Object3D): void {
+  const skinnedMeshes = new Set<THREE.SkinnedMesh>();
+  const geometryToSkinnedMesh = new Map<THREE.BufferGeometry, THREE.SkinnedMesh>();
+
+  // Traverse entire tree and collect skinned meshes
+  root.traverse((obj) => {
+    if (obj.type !== 'SkinnedMesh') {
+      return;
+    }
+
+    const skinnedMesh = obj as THREE.SkinnedMesh;
+
+    // Check if the geometry has already been encountered
+    const previousSkinnedMesh = geometryToSkinnedMesh.get(skinnedMesh.geometry);
+    if (previousSkinnedMesh) {
+      // Skinned meshes that share their geometry with other skinned meshes can't be processed.
+      // The skinnedMeshes already contain previousSkinnedMesh, so remove it now.
+      skinnedMeshes.delete(previousSkinnedMesh);
```

Comment:
I doubt this code won't work correctly if we encounter an odd count (3, 5, ...) of skinned meshes that shares a same geometry? 50% confidence

Reply:
Why we can't process such meshes by the way? Is it because we are going to modify the geometry skinIndex to match to the reorganized skeleton?

Reply:
Actually it should work reliably, as the mesh is removed from `skinnedMeshes`, but still in `geometryToSkinnedMesh`, which is used to check if the geometry has been seen before. So in case of 3 `SkinnedMeshes` with the same geometry, you'd get:

1. `previousSkinnedMesh = undefined` => `skinnedMeshes.add(skinnedMesh)`
2.  `previousSkinnedMesh = <first mesh>` => `skinnedMeshses.delete(previousMesh)`
3.  `previousSkinnedMesh = <first mesh>` => `skinnedMeshses.delete(previousMesh)` (no-op)

> Why we can't process such meshes by the way? Is it because we are going to modify the geometry skinIndex to match to the reorganized skeleton?

Correct, the `skinIndex` will be changed which would cause issues when processing the next `SkinnedMesh` with that geometry. The problem gets even worse when the geometry is the same, but the skeleton is different. Neither cases are likely and I have yet to see any model with such a setup, but I thought it would be better to have this check as a safeguard, rather than to assume geometries aren't shared.

Reply:
> Actually it should work reliably, as the mesh is removed from `skinnedMeshes`, but still in `geometryToSkinnedMesh`, which is used to check if the geometry has been seen before.

Ah, I simply missaw that the code deletes the item from `skinnedMeshes`, not `geometryToSkinnedMesh`. Thank you for clarifying.